File: smplwise_vms/backend/smplwise/routers/health.py
from __future__ import annotations

import os
import sqlite3
from typing import Any

from fastapi import APIRouter, Depends, Request

from .. import __version__
from ..auth import current_principal, get_conn, settings_of
from ..db import permission_revision, unlocked
from ..rbac import INSTALLATION, Principal, require
from ..services import autosync, events_derive, events_ingest, ha_client, ha_sync
from ..services import health_report as health_report_svc

router = APIRouter()


@router.get("/health")
def health(request: Request, principal: Principal = Depends(current_principal), conn: sqlite3.Connection = Depends(get_conn)) -> dict[str, Any]:
    settings = settings_of(request)
    return {
        "status": "ok",
        "version": __version__,
        "db": {"ok": conn.execute("SELECT 1").fetchone()[0] == 1, "permission_revision": permission_revision(conn)},
        "data_dir_writable": os.access(settings.data_dir, os.W_OK),
        "nvr_configured": bool(settings.nvr_host and settings.nvr_user),
        "go2rtc_configured": bool(settings.go2rtc_url),
        "discovery": {**autosync.STATE, "cameras": conn.execute("SELECT COUNT(*) FROM cameras").fetchone()[0], "interval_s": autosync.INTERVAL_S},
        "events": {"ingest": events_ingest.STATE.as_dict(), "derive": events_derive.STATE, "stored": conn.execute("SELECT COUNT(*) FROM events").fetchone()[0]},
        "home_assistant": {"configured": ha_client.configured(settings), **ha_sync.STATE.as_dict()},
        "identity_source": principal.source,
        "renderer": "pdftoppm" if any(os.access(os.path.join(p, "pdftoppm"), os.X_OK) for p in os.environ.get("PATH", "").split(os.pathsep)) else "pymupdf-or-none",
    }


@router.get("/health/report")
def health_report(request: Request, fresh: bool = False, principal: Principal = Depends(current_principal), conn: sqlite3.Connection = Depends(get_conn)) -> dict[str, Any]:
    """One status per subsystem (T033). Device probes are network calls, so they run outside the request's
    write transaction; `fresh=1` drops the probe cache first."""
    require(conn, principal, "system.configure", INSTALLATION)
    if fresh:
        health_report_svc.invalidate()
    settings = settings_of(request)
    with unlocked(conn):
        report = health_report_svc.build(settings, conn, probe=True)
    return report


@router.get("/health/summary")
def health_summary(request: Request, principal: Principal = Depends(current_principal), conn: sqlite3.Connection = Depends(get_conn)) -> dict[str, Any]:
    """What the top bar shows every signed-in user: cached states only, no device probes, no details that a
    non-administrator should not see (labels are operator wording, never addresses or credentials)."""
    return health_report_svc.summary(settings_of(request), conn)

File: frontend/src/api/health.ts
/** Health report (T033): one check per subsystem with its own status, detail and numbers. */
import { get } from './client';

export type CheckStatus = 'ok' | 'warn' | 'error';

export interface HealthCheck {
  id: string;
  label: string;
  status: CheckStatus;
  detail: string;
  meta: Record<string, unknown>;
}

export interface HealthReport {
  status: CheckStatus;
  version: string;
  uptime_s: number;
  checked_at: string;
  probe_ttl_s: number;
  checks: HealthCheck[];
}

export const STATUS_LABEL: Record<CheckStatus, string> = { ok: 'תקין', warn: 'לתשומת לב', error: 'שגיאה' };
export const STATUS_KIND: Record<CheckStatus, 'live' | 'stale' | 'offline'> = { ok: 'live', warn: 'stale', error: 'offline' };

export const healthReport = (fresh = false) => get<HealthReport>(`health/report${fresh ? '?fresh=1' : ''}`);

/** Cheap status for the top bar (every signed-in user): cached job states only, never a device probe. */
export interface HealthSummary {
  status: CheckStatus;
  items: { id: string; status: CheckStatus; label: string }[];
  checked_at: string;
  version: string;
}
export const healthSummary = () => get<HealthSummary>('health/summary');

export function fmtUptime(s: number): string {
  if (s < 3600) return `${Math.floor(s / 60)} דק׳`;
  if (s < 86400) return `${Math.floor(s / 3600)} שע׳ ${Math.floor((s % 3600) / 60)} דק׳`;
  return `${Math.floor(s / 86400)} ימים ${Math.floor((s % 86400) / 3600)} שע׳`;
}
